Sources/XCTest/XCTestCase.swift

```diff
@@ -8,12 +8,58 @@
 //
 //
 //  XCTestCase.swift
-//  Base protocol (and extension with default methods) for test cases
+//  Base class for test cases
 //
 
-public protocol XCTestCase : XCTestCaseProvider {
-    func setUp()
-    func tearDown()
+public typealias XCTestMethod = XCTestCase throws -> Void
+
+public enum XCTestError: ErrorType, CustomStringConvertible {
+    case MissingAllTestsOverride
+
+    public var description: String {
+        switch self {
+        case .MissingAllTestsOverride: return "This test case class must override the allTests property"
+        }
+    }
+}
+
+public class XCTestCase {
+
+    public required init() {
+    }
+
+    public func setUp() {
+    }
+
+    public func tearDown() {
+    }
+
+    /// In the Objective-C version of XCTest, it is possible to discover all tests when the test is executed
+    /// by asking the runtime for all methods and looking for the string "test". In Swift, we ask test case subclasses
+    /// to tell us the list of tests by implementing this property. This should look like:
+    ///
+    ///     override class var allTests : [(String, XCTestMethod)] {
+    ///         return [
+    ///             ("test_foo", test(test_foo)),
+    ///             ("test_bar", test(test_bar)),
+    ///         ]
+    ///     }
+    public class var allTests: [(String, XCTestMethod)] {
+        return [("missingAllTestsOverride", { _ in throw XCTestError.MissingAllTestsOverride })]
```

Comment:
Ingenious! I was expecting to see the base implementation call `fatalError()`, but this approach is much better. :+1: